Restore a very narrow function argument conversion for -swift-version 4.

rudkx opened the pull request:
Allow functions with type `(()) -> T` to be passed in places where we
expect `() -> T`, but only for -swift-version 4 (for -swift-version 3
this already works due to other horrible things in CSSimplify.cpp).

We need to look at how we can help migrate these cases to
-swift-version 5, but in the meantime, but that is something we can
consider separately.

Commit:
Restore a very narrow function argument conversion for -swift-version 4.

Allow functions with type `(()) -> T` to be passed in places where we
expect `() -> T`, but only for -swift-version 4 (for -swift-version 3
this already works due to other horrible things in CSSimplify.cpp).

We need to look at how we can help migrate these cases to
-swift-version 5, but in the meantime, but that is something we can
consider separately.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

xedin reviewed (approved):
LGTM 😭

jrose-apple commented on test/Constraints/tuple_arguments.swift:
```diff
@@ -1626,7 +1626,7 @@ func rdar33239714() {
   func foo(_: (() -> Void)?) {}
   func bar() -> ((()) -> Void)? { return nil }
-  foo(bar()) // expected-error {{cannot convert value of type '((()) -> Void)?' to expected argument type '(() -> Void)?'}}
+  foo(bar()) // Allow ((()) -> Void)? to be passed in place of (() -> Void)? for -swift-version 4 but not later.
```
Can you make a Swift 5 test that shows it failing?

rudkx replied:
Yes. I considered that earlier today, wondering what the “right time” would be to fork this test again for Swift 5. Perhaps now is that time.

Commit:
Fork tuple_arguments test for Swift 5.

Move the existing test into Compatibility/tuple_arguments_4.swift.
Move the Swift 3 test, Compatibility/tuple_arguments.swift, into
Compatibility/tuple_arguments_4.swift.

Add a couple new tests to each file to capture what we do in each
mode.

Overall this doesn't seem like the most satisfying approach for
covering compatibility testing in our standard commit testing, but
it's all we've got at the moment.

rudkx commented:
@swift-ci Please smoke test

NachoSoto commented:
So when Swift 5 comes out we're gonna have to file regression bug again?

rudkx commented:
No, the code will need to change for `-swift-version 5`.

It was a bug that Swift ever allowed lack of a parameter to be conflated with having a `Void` (aka `()`) parameter. This bug arose because of a representation problem in the compiler. Some effort has gone into fixing that problem, resulting in code that previously compiled now failing to compile. The tuple-splatting behavior was another symptom of this same representation problem. In fact you could consider this another instance of the splatting behavior - the zero-element tuple to no-argument case.

For the particular code for which this problem was reported, I would think overloading `log` (and `logIfNeeded`) to return a no-parameter function would be sufficient to allow the code to compile in both `-swift-version 4` mode and `-swift-version 5` mode.

This works in a stand-alone test case, and has the benefit that you can remove the `as ((()) -> Void)?` coercions:

```
func logIfNeeded<T>() -> ((T) -> Void)? { fatalError() }
func logIfNeeded() -> (() -> Void)? { fatalError() }

func log<T>() -> ((T) -> Void)? { return logIfNeeded() }
func log() -> (() -> Void)? { return logIfNeeded() }

func on(no: (()->())? = nil, yes: ((())->())? = nil) {}

on(no: log(), yes: log())
```

NachoSoto commented:
> No, the code will need to change for -swift-version 5.

Okay, but right now there is no way to make that code compile, as I pointed out in [the Jira](https://bugs.swift.org/browse/SR-6796).

> For the particular code for which this problem was reported, I would think overloading log (and logIfNeeded) to return a no-parameter function would be sufficient to allow the code to compile in both -swift-version 4 mode and -swift-version 5 mode.

I'll give that a try, thanks. In my opinion, however, [the Jira](https://bugs.swift.org/browse/SR-6796) shouldn't be closed because, whatever the Swift decides about whether that code should compile or not, `error: generic parameter 'T' could not be inferred` is not a useful diagnostic for users to understand that the code is hitting a limitation of the language and what the workaround should be.

Also, I want to point out that [the reason why `ReactiveCocoa` (a superset of `ReactiveSwift`) was disabled from the source-compatibility suite [is still not fixed](https://bugs.swift.org/browse/SR-6658), so because of that, not only did it not catch this regression, but it's potentially still not being able to catch future ones.

NachoSoto commented:
@rudkx that workaround doesn't work:
<img width="739" alt="screen shot 2018-02-08 at 10 45 11 am" src="https://user-images.githubusercontent.com/685609/35991792-2870705c-0cbd-11e8-91ce-e81b58f41353.png">

NachoSoto commented:
IMO, if this could work:
<img width="544" alt="screen shot 2018-02-08 at 10 49 38 am" src="https://user-images.githubusercontent.com/685609/35991981-c996eb1e-0cbd-11e8-99e5-a43ba4cad9af.png">

Then we wouldn't be having arguments about `()` and `Void`. The problem is that `T` can't be inferred by itself, and there's no way to tell the compiler what type to use for this case.

rudkx commented:
> @rudkx that workaround doesn't work:

Does the stand-alone test case that I pasted in compile for you?

What compiler are you using when you build and hit the error about the conflicting definition?

NachoSoto commented:
Your example does compile, but not the full example as shown in my screenshot. I welcome you to clone the repo as explained in [SR-6796](https://bugs.swift.org/browse/SR-6796) and you can see by yourself.

rudkx commented:
Apparently we do not allow overloading nested functions, so I can only offer this slightly cumbersome work-around:

```
extension Signal {
  struct LogContext {
    let logger: EventLogger
    let events: Set<LoggingEvent.Signal>
    let identifier: String
    let fileName: String
    let functionName: String
    let lineNumber: Int

    func log<T>(_ event: LoggingEvent.Signal) -> ((T) -> Void)? {
      return event.logIfNeeded(events: events) { event in
	self.logger(self.identifier, event, self.fileName, self.functionName, self.lineNumber)
      }
    }

    func log(_ event: LoggingEvent.Signal) -> (() -> Void)? {
      return event.logIfNeeded(events: events) { event in
	self.logger(self.identifier, event, self.fileName, self.functionName, self.lineNumber)
      }
    }
  }

...
  public func logEvents(identifier: String = "", events: Set<LoggingEvent.Signal> = LoggingEvent.Signal.allEvents, fileName: String = #file, functionName: String = #function, lineNumber: Int = #line, logger: @escaping EventLogger = defaultEventLog) -> Signal<Value, Error> {
    let logContext =
    LogContext(logger: logger,
      events: events,
      identifier: identifier,
      fileName: fileName,
      functionName: functionName,
      lineNumber: lineNumber)

    return self.on(
      failed: logContext.log(.failed),
      completed: logContext.log(.completed),
      interrupted: logContext.log(.interrupted),
      terminated: logContext.log(.terminated),
      disposed: logContext.log(.disposed),
      value: logContext.log(.value)
    )
  }
```

rudkx commented:
> IMO, if this could work:

I would love to make that work, but it's not possible today without completely fixing the function type representation. Changes make that work will also almost certainly break other code as well.

NachoSoto commented:
I'm trying your workaround, it compiles now, but tests aren't passing so I'm investigating.
In the mean time I want to say that this still smells like a huge regression. Why would we need an overload to a generic function for `T` = `Void`? `Void` (at least from the user's standpoint) is a type like any other, so this is as arbitrary as if the type checker wasn't able to resolve the usage of a generic function for `String` and we needed an overload for that... what a huge step backwards :(

NachoSoto commented:
But hey at least we're getting Python interpretability, so yay? 🤦‍♂️

jrose-apple commented:
Void is indeed a type: it's a typealias for the empty tuple. A function type that takes Void is written `(Void) -> Int`, or without sugar `(()) -> Int`, and is called by passing a value of type Void to the function: `fn(())`. A function that takes no arguments is *not* a function that takes Void (one argument).

Past compiler releases *mistakenly* treated `() -> Int` and `(()) -> Int` as interchangeable in some cases, but "from the user's standpoint" it never actually made sense. But I'll acknowledge a user probably wouldn't bother thinking about it if the compiler doesn't complain, hence the use in ReactiveSwift.

NachoSoto commented:
@jrose-apple if I'm reading that correctly, sounds like you agree with me about the fact that we shouldn't require an overload for `Void`, right?

Can we count on a fix for this for Swift 4.1?

jrose-apple commented:
No, I'm saying the opposite. You don't need an overload for Void; you need an overload for "no arguments".

NachoSoto commented:
I see what you're saying. Why can't the compiler correctly convert between `(Void)` and `() -> Void` though?

I'm looking into this test failure with the workaround that @rudkx proposed, and there's something really weird going on. It compiles, but the test (for this operator) failure seems to indicate there's a weird miscomputation going on. Still digging...

jrose-apple commented:
I'm not sure what you mean by "correctly"; they are different types and they should be different types. Of course we (inclusive "we", the Swift community, possibly via the Evolution process) *could* formally introduce a conversion, but conversions between unrelated types are the kind of thing that makes type checking slow, especially when the conversion goes both ways. We'd also either have to put this into dynamic casting as well, or have different static and dynamic behavior.

```swift
func fn<T>(_ arg: T) -> Void {
  print(arg)
}

let boxedFn: Any = fn as (Void) -> Void
print(boxedFn is (Void) -> Void) // true
print(boxedFn is () -> Void) // false
```

NachoSoto commented:
I mean, you tell me. This is what I had to do to get @rudkx's workaround working:

```swift
private extension LoggingEventProtocol {
	// See https://bugs.swift.org/browse/SR-6796.
	func logIfNeeded(events: Set<Self>, logger: @escaping (String) -> Void) -> (() -> Void)? {
		return (self.logIfNeeded(events: events, logger: logger) as ((()) -> Void)?)
			.map { closure in
				{ closure(()) }
			}
	}
	
	func logIfNeeded<T>(events: Set<Self>, logger: @escaping (String) -> Void) -> ((T) -> Void)? {
		guard events.contains(self) else {
			return nil
		}
		
		return { value in
			if value is Void {
				logger("\(self.rawValue)")
			} else {
				logger("\(self.rawValue) \(value)")
			}
		}
	}
}
```

Can you really tell me with a straight face that this regression is actually _a feature_ for `Swift 5.0`? We could spend hours discussing the philosophy behind `Void` and `()`, but at the end of the day this change means we have to explicitly do this:
```swift
(closure as ((()) -> Void)?)
    .map { closure in
        { closure(()) }
    }
```

Which is an absolute nightmare, and with the current state of diagnostics in `Swift` I guarantee you most people won't be able to understand the errors behind this.

rudkx commented:
FYI, I have been busy this afternoon but this morning I wrote up a pitch for fixing this that I will post to the Evolution forum and link here, hopefully tonight.

NachoSoto commented:
👌

jrose-apple commented:
IMO the right answer was *always* "pass `()` when you call a function whose generic parameter is Void". It's a mistake that people got used to writing `f()` when the language wanted them to write `f(())`.

(Is the latter *better?* No, I'm not going to claim that. But it's *consistent,* and would have meant a faster compiler with fewer bugs and potentially better diagnostics.)

rudkx commented:
> FYI, I have been busy this afternoon but this morning I wrote up a pitch for fixing this that I will post to the Evolution forum and link here, hopefully tonight.

See: https://forums.swift.org/t/more-consistent-function-types/9765
